### PR TITLE
Resolution: Preparing the Raku Ecosystem for the Future

### PR DESCRIPTION
**Pull request by lizmat**

### CODE REVIEW

File: solutions/meta/sunsetting-p6c-cpan.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+- Migrate the remaining raku-community-modules to zef before July 1st 2022
+There are about 95 modules in the raku-community-modules repo that still need migrating to zef ecosystem.  More than 20 have already been moved.  Elizabeth Mattijsen intends to move them all by July 1st, unless someone beats her to it.
+
+- Alert any module authors after July 1st, 2022
+Any module authors still not having moved their modules to the zef ecosystem by July 1st 2022, should be informed with a i**single** issue that lists the modules in question.  This could be an issue in the problem solving repo, but another repo would work as well.  The module authors should also be able to indicate they no longer wish to support a module themselves: in that case, a move to the raku-community-modules repo should be considered.
+
+- Stop the ecosystem harvesters on January 1st, 2023
+Stop running the ecosystem harvesters of p6c and CPAN.  This will effectively disregard any updates of any modules on those ecosystems.  But any updates with a new version would still become visible in the REA, as that runs its own harvester.
```
Comment (ugexe) [inline solutions/meta/sunsetting-p6c-cpan.md:22]: If REA continues to run its own harvester then the security issues of e.g. putting zef ecosystem stuff in a non zef ecosystem

Comment (lizmat) [reply solutions/meta/sunsetting-p6c-cpan.md:22]: So you're saying the REA harvester should also be switched off?

Comment (japhb) [reply solutions/meta/sunsetting-p6c-cpan.md:22]: @ugexe You're right, but are you arguing for turning off all harvesters simultaneously, or shortening the period in which only the REA harvester remains active, or just saying that this doc should make clear that the security problem isn't fixed until all harvesters are shut off?  (Or have I missed an option?)

Comment (ugexe) [reply solutions/meta/sunsetting-p6c-cpan.md:22]: Let me rephrase, as my comment was actually more related to being enabled by default (I typed it on a phone and got cut short):

Generally we should be moving away from opt-in security (i.e. having to disable ecosystems that don't enforce `auth`) to opt-out security (i.e. having to enable ecosystems that don't enforce `auth`). In that respect it could be argued REA shouldn't default to enabled in zef, but indeed could continue to run its harvester. In other words -- we probably want REA being enabled as a default to be a temporary thing and not give an impression that would encourage people to continue using them over the secure alternative(s)

Comment (japhb) [reply solutions/meta/sunsetting-p6c-cpan.md:22]: Ah, I think I understand -- you'd like a sunset date for REA staying default-enabled?

I would say "Why don't we just pick a date for REA to stop serving CPAN- and p6c-harvested content?"  But then I realized that one of the PROs for Zef is that it never forgets *anyway*.

Which then brought up the question: "What do we do about legal take-downs or suchlike?"  But again, REA doesn't ameliorate that problem much.

So I guess a sunset data for REA default-enable makes sense to me too.

File: solutions/meta/sunsetting-p6c-cpan.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Sunsetting p6c / CPAN ecosystems
+
+This document provides a solution to [issue 316](https://github.com/Raku/problem-solving/issues/316).
+
+## Announcement
+
+The p6c and CPAN ecosystems are to be sunsetted as far as installation of Raku modules is concerned.  This will consist of the following steps:
```
Comment (JJ) [inline solutions/meta/sunsetting-p6c-cpan.md:7]: alt: deprecated instead of "sunsetted"

File: solutions/meta/sunsetting-p6c-cpan.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+This will allow any installation of a Raku module to look through the [Raku Ecosystem Archive (REA)](https://github.com/raku/REA) if not found on zef, p6c or CPAN ecosystems.
+
+- Disable the p6c / CPAN ecosystems by default in zef immediately
+With the REA enabled, the modules that currently reside on p6c and CPAN ecosystems can be downloaded and installed from the REA.
+
+- Migrate the remaining raku-community-modules to zef before July 1st 2022
+There are about 95 modules in the raku-community-modules repo that still need migrating to zef ecosystem.  More than 20 have already been moved.  Elizabeth Mattijsen intends to move them all by July 1st, unless someone beats her to it.
+
+- Alert any module authors after July 1st, 2022
+Any module authors still not having moved their modules to the zef ecosystem by July 1st 2022, should be informed with a i**single** issue that lists the modules in question.  This could be an issue in the problem solving repo, but another repo would work as well.  The module authors should also be able to indicate they no longer wish to support a module themselves: in that case, a move to the raku-community-modules repo should be considered.
```
Comment (JJ) [inline solutions/meta/sunsetting-p6c-cpan.md:19]: scratch "i"

**Review by JJ — DISMISSED** [review]
OK, except for minor nits.

**Review by JJ — APPROVED** [review]
:+1:

**Review by japhb — APPROVED** [review]

**Review by jonathanstowe — APPROVED** [review]

**Review by codesections — APPROVED** [review]

File: solutions/meta/sunsetting-p6c-cpan.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+This will allow any installation of a Raku module to look through the [Raku Ecosystem Archive (REA)](https://github.com/raku/REA) if not found on zef, p6c or CPAN ecosystems.
+
+- Disable the p6c / CPAN ecosystems by default in zef immediately
+With the REA enabled, the modules that currently reside on p6c and CPAN ecosystems can be downloaded and installed from the REA.
+
+- Migrate the remaining raku-community-modules to zef before July 1st 2022
+There are about 95 modules in the raku-community-modules repo that still need migrating to zef ecosystem.  More than 20 have already been moved.  Elizabeth Mattijsen intends to move them all by July 1st, unless someone beats her to it.
+
+- Alert any module authors after July 1st, 2022
+Any module authors still not having moved their modules to the zef ecosystem by July 1st 2022, should be informed with a **single** issue that lists the modules in question.  This could be an issue in the problem solving repo, but another repo would work as well.  The module authors should also be able to indicate they no longer wish to support a module themselves: in that case, a move to the raku-community-modules repo should be considered.
```
Comment (Altai-man) [inline solutions/meta/sunsetting-p6c-cpan.md:19]: I'm a bit afraid of this step. We have more than 1500 modules right now, I have a vague feeling most of them are unmaintained for various reasons, and for numerous ones authors are not available at all (even to indicate it). We also don't have a huge queue of adopters, so this potentially floods the raku-community-modules into chaos.

Comment (Altai-man) [reply solutions/meta/sunsetting-p6c-cpan.md:19]: And speaking of the last two steps, again too much burden is put onto raku-community-modules which is already not very thriving on its own iiuc.

Comment (lizmat) [reply solutions/meta/sunsetting-p6c-cpan.md:19]: > We also don't have a huge queue of adopters, so this potentially floods the raku-community-modules into chaos.

True, but about 1/4 of modules have already been moved, and good example (hopefully) gets good following.  FWIW, if it hadn't been for the comments I got from adopting `IRC::Client`, I would have adopted many more modules into "my own" set of modules.

> And speaking of the last two steps, again too much burden is put onto raku-community-modules which is already not very thriving on its own iiuc.

Well, raku-community-modules could definitely get some TLC.  But I'm willing to give that (and have been giving that in the past months already), as an up-to-date ecosystem is vital for the Raku Programming Language.

Well-kept and well-documented modules will make it easier for people to submit PR's or maintain themselves.

Comment (jonathanstowe) [reply solutions/meta/sunsetting-p6c-cpan.md:19]: > don't have a huge queue of adopters

FWIW I'm probably up for adopting anything that is genuinely useful where I understand the problem domain.

Comment (Altai-man) [reply solutions/meta/sunsetting-p6c-cpan.md:19]: Fair enough.

**Review by Altai-man — APPROVED** [review]

**Review by CIAvash — APPROVED** [review]

**Comment by lizmat** [comment]
image

Something tells me this is never going to work this way  :-(

**Comment by patrickbkr** [comment]
@lizmat We are at day 8 of 14. I do agree though that the process feels pointless if only in theory solutions are accepted via approvals. Maybe a change of perspective helps. A proposal needs to survive 2 weeks without any disprovals. (Given there is reason to believe people actually looked at it.)

**Comment by lizmat** [comment]
My point was really about 8 + 4 < 13  so the requirement as it is currently stated, can **never** be reached.

**Comment by patrickbkr** [comment]
:face_palm:

**Review by FCO — APPROVED** [review]

**Review by japhb — APPROVED** [review]
Don't know why my previous approval was reverted, but approving again since it seems to think I didn't.

**Review by JJ — APPROVED** [review]
:-1:

**Review by Kaiepi — APPROVED** [review]

**Review by jonathanstowe — APPROVED** [review]

**Review by masak — APPROVED** [review]

**Comment by lizmat** [comment]
Here we are, almost a month later.

This process isn't working.  And Github apparently cannot keep track either: in the sidebar it shows 11 approvals, yet just above this message it says 8 approvals.

The 8 remaining people who I asked for a review, either are too lazy to do a review, or are not in agreement and don't want to give an approval: there is no way to know.

I'm torn between just merging the PR and get on with it, or just closing the PR and let everything ecosystem wise go to hell.

**Comment by niner** [comment]
I don't understand why Github requires 13 approvals for this PR, when https://github.com/Raku/problem-solving only lists 12 designated reviewers. Is this some wrong setting in the repository?

Also the rules say: "A consensus does not require unanimity, but it should be clear that the Raku community as a whole supports the solution. In particular, if 14 days pass between the time when the PR (or the latest revision of the PR, if it's edited) was proposed and none of the reviewers (listed below) have objected, then the proposer of the PR can use their judgement to determine whether consensus exists."

Since there are 11 approvals and no objections whatsoever, if I were the proposer, I'd judge this as consensus.

Note that I didn't review, because I'm not one of the designated reviewers in problem-solving. However, I do support this proposal.

**Comment by lizmat** [comment]
Not sure what's going on here, but Github **currently** allows setting of minimal approving reviews to **6** maximally.  So *12* appears out of the question.

**Comment by lizmat** [comment]
> none of the reviewers (listed below) have objected

That's the thing: @ugexe had some remarks, after which I made changes and asked for a review again.  And he didn't approve after that.  So in that sense, the PR is in limbo.

**Comment by coke** [comment]
the list of approvers inline in comments doesn't show @japhb as having approved or being pending, despite his approval on the list on the right; I wonder if this has to do with his permissions on the repo?

**Comment by coke** [comment]
Again on @japhb  - note his approval checkmark is styled ".color-fg-muted", but mine is ".color-fg-success" which is a very subtle difference to my eye.

**Review by alabamenhu — APPROVED** [review]

**Review by vrurg — APPROVED** [review]

**Comment by vrurg** [comment]
The difference in coloring is really so subtle that it's only now as I've noticed that mine checkmark was not green.

Anyway, I consider the 14 days rule clearly apply here as there are no disapprovals.

**Review by jonathanstowe — APPROVED** [review]

**Comment by jnthn** [comment]
> too lazy to do a review

That's not the most charitable spelling of "too over-burdened with other things to have time to do a meaningful review". :-) I don't want to be in the way of anything, but note that I'm already [not on the reviewers list here](https://github.com/Raku/problem-solving#reviewers); is there somewhere else I should remove myself? (I'll still comment here as time allows, but my time for Raku is quite limited these days, and I'd rather focus the time I do have on the RakuAST work, getting EA into MoarVM, etc.)

**Comment by lizmat** [comment]
@jnthn point taken, my apologies.

Re asking you for a review: normally I wouldn't have, but since this has very wide implications, I asked specifically for your review.  It won't happen again :)

**Comment by lizmat** [comment]
I will merge tomorrow when I'm less tired  :-)

**Comment by lizmat** [comment]
Thank you everybody for your cooperation!
